From 4fe63111e3cc9263c73cd0eefb8a1440c51f3ab9 Mon Sep 17 00:00:00 2001
From: Dmitry Kasatkin <d.kasatkin@samsung.com>
Date: Wed, 29 Oct 2014 11:17:50 +0200
Subject: [PATCH 14/45] ima: provide appraise_type=evmsig

Extended attributes provide means to assign capabilities to the processes
and security labels. System might want force to require to have them
immutable and protected  by signature.

This patch provides 'appraise_type=evmsig' option to require EVM signatures.

Signed-off-by: Dmitry Kasatkin <d.kasatkin@samsung.com>
---
 security/integrity/evm/evm_main.c   | 6 ++++++
 security/integrity/ima/ima_policy.c | 2 ++
 security/integrity/integrity.h      | 1 +
 3 files changed, 9 insertions(+)

diff --git a/security/integrity/evm/evm_main.c b/security/integrity/evm/evm_main.c
index 46b65ac..0666f6d 100644
--- a/security/integrity/evm/evm_main.c
+++ b/security/integrity/evm/evm_main.c
@@ -153,6 +153,10 @@ static enum integrity_status evm_verify_hmac(struct dentry *dentry,
 	/* check value type */
 	switch (xattr_data->type) {
 	case EVM_XATTR_HMAC:
+		if (iint && (iint->flags & EVM_DIGSIG_REQUIRED)) {
+			rc = -EACCES;
+			break;
+		}
 		rc = evm_calc_hmac_or_hash(dentry, xattr_name, xattr_value,
 					   xattr_value_len, EVM_XATTR_HMAC,
 					   calc.digest);
@@ -165,6 +169,8 @@ static enum integrity_status evm_verify_hmac(struct dentry *dentry,
 		break;
 	case EVM_IMA_XATTR_DIGSIG:
 		version = ((const char *)xattr_data)[1];
+		if (iint)
+			set_bit(EVM_DIGSIG, &iint->atomic_flags);
 		rc = evm_calc_hmac_or_hash(dentry, xattr_name, xattr_value,
 					   xattr_value_len,
 					   version == 3 ? EVM_IMA_XATTR_DIGSIG :
diff --git a/security/integrity/ima/ima_policy.c b/security/integrity/ima/ima_policy.c
index c9944ca..d558031 100644
--- a/security/integrity/ima/ima_policy.c
+++ b/security/integrity/ima/ima_policy.c
@@ -648,6 +648,8 @@ static int ima_parse_rule(char *rule, struct ima_rule_entry *entry)
 			ima_log_string(ab, "appraise_type", args[0].from);
 			if ((strcmp(args[0].from, "imasig")) == 0)
 				entry->flags |= IMA_DIGSIG_REQUIRED;
+			else if ((strcmp(args[0].from, "evmsig")) == 0)
+				entry->flags |= EVM_DIGSIG_REQUIRED;
 			else
 				result = -EINVAL;
 			break;
diff --git a/security/integrity/integrity.h b/security/integrity/integrity.h
index 76786ef..c55c015 100644
--- a/security/integrity/integrity.h
+++ b/security/integrity/integrity.h
@@ -31,6 +31,7 @@
 #define IMA_DIGSIG_REQUIRED	0x01000000
 #define IMA_PERMIT_DIRECTIO	0x02000000
 #define IMA_NEW_FILE		0x04000000
+#define EVM_DIGSIG_REQUIRED	0x08000000
 
 #define IMA_DO_MASK		(IMA_MEASURE | IMA_APPRAISE | IMA_AUDIT | \
 				 IMA_APPRAISE_SUBMASK)
-- 
2.1.4

